### PR TITLE
Add support for navigating to the definition of the target label of a `goto`

### LINKED ISSUE #69916: Support go to definition (F12) for `goto` statements, navigating to the target label
## Summary

Support go to definition (F12) for `goto` statements, navigating to the target label

## Background and Motivation

In switch statements more often, and perhaps in other parts of the code, `goto` statements could benefit a lot from this feature, by having a direct way to navigate to the target label that is referenced in the code, offering a better experience reading and understanding the logic.

## Proposed Feature

The proposal is to allow pressing `F12` on the `goto` statement, or the `default` keyword for a switch statement label, or the `case` keyword for the switch statement label. Examples of acceptable navigation include (with markup):
```csharp
public void Method(int argument)
{
    switch (argument)
    {
        case 1:
            [goto] [default];
        case 2:
            [goto] [case] 1;
        default:
            break;
    }
}
```

If the user uses the go to definition feature on the spans within `[]`, then they are navigated to the target label that is referenced. In the case of `goto default`, the navigator navigates to the `default:` label within the switch statement, and the `goto case 1` statement would navigate to the `case 1:` label within the switch statement.

## Alternative Designs

None.

### PR DESCRIPTION
**Pull request by Rekkonnect**
Closes #69916 

Adds support for both C# and VB. Using go to navigation on the keywords `goto`, `case` and `default` in a go to statement will navigate to the definition of the target label.

### CODE REVIEW

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs [hunk 1]
```diff
@@ -65,6 +68,25 @@ protected override ISymbol FindRelatedExplicitlyDeclaredSymbol(ISymbol symbol, C
...
+                        if (foundAccessibleLabel is null)
+                        {
+                            return null;
+                        }
+
+                        var symbol = semanticModel.GetDeclaredSymbol(foundAccessibleLabel);
+                        if (symbol is null)
+                        {
+                            return null;
+                        }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:88]:
```suggestion
                            return null;

                        var symbol = semanticModel.GetDeclaredSymbol(foundAccessibleLabel);
                        if (symbol is null)
                            return null;
```

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs [hunk 2]
```diff
@@ -119,6 +136,38 @@ protected override ISymbol FindRelatedExplicitlyDeclaredSymbol(ISymbol symbol, C
...
                 return node;
             }
+
+            SyntaxNode? TryFindAccessibleLabel(SyntaxNode? node)
+            {
+                var statement = node?.FirstAncestorOrSelf<GotoStatementSyntax>();
+                if (statement is null)
+                {
+                    return null;
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:146]:
```suggestion
                if (statement is null)
                    return null;
```

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs [hunk 3]
```diff
@@ -119,6 +136,38 @@ protected override ISymbol FindRelatedExplicitlyDeclaredSymbol(ISymbol symbol, C
...
+
+                var gotoOperation = semanticModel.GetOperation(statement) as IBranchOperation;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:148]: 'as' checks have to be null checked.

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs [hunk 4]
```diff
@@ -119,6 +136,38 @@ protected override ISymbol FindRelatedExplicitlyDeclaredSymbol(ISymbol symbol, C
 
                 return node;
             }
+
+            SyntaxNode? TryFindAccessibleLabel(SyntaxNode? node)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:140]:
```suggestion
            SyntaxNode? TryFindAccessibleLabel(SyntaxNode node)
```

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs [hunk 5]
```diff
@@ -119,6 +136,38 @@ protected override ISymbol FindRelatedExplicitlyDeclaredSymbol(ISymbol symbol, C
...
+                if (statement is null)
+                {
+                    return null;
+                }
+
+                var gotoOperation = semanticModel.GetOperation(statement) as IBranchOperation;
+                Debug.Assert(gotoOperation is { BranchKind: BranchKind.GoTo });
+                var target = gotoOperation.Target;
+
+                var expression = statement.Expression;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:152]: i dont' understand why the rest of this is needed.  if you have the target, isn't that an ILabelSymbol?  Why not just return its location?

File: src/Features/VisualBasic/Portable/GoToDefinition/VisualBasicGoToDefinitionSymbolService.vb [hunk 6]
```diff
@@ -48,6 +49,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.GoToDefinition
...
+                Return labelTarget.GetFirstToken().Span.Start
+            End If
+
+            Return Nothing
+        End Function
+
+        Private Shared Function TryGetLabelTarget(semanticModel As SemanticModel, node As SyntaxNode) As SyntaxNode
+            Select Case node.Kind()
+                Case SyntaxKind.GoToStatement
+                    Dim goToStatement = DirectCast(node, GoToStatementSyntax)
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/GoToDefinition/VisualBasicGoToDefinitionSymbolService.vb:63]: why is this code not using IOperation?

**Comment by CyrusNajmabadi** [comment]
Done with pass.

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs (hunk 5 above)
Comment (Rekkonnect) [reply src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:152]: Correct observation, there is no reason for the additional processing; the target is calculated and stored within the IOperation

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs [hunk 7]
```diff
@@ -119,6 +137,20 @@ protected override ISymbol FindRelatedExplicitlyDeclaredSymbol(ISymbol symbol, C
...
+                var statement = node.FirstAncestorOrSelf<GotoStatementSyntax>();
+                if (statement is null)
+                    return null;
+
+                if (semanticModel.GetOperation(statement) is not IBranchOperation gotoOperation)
+                    return null;
+
+                Debug.Assert(gotoOperation is { BranchKind: BranchKind.GoTo });
+                var target = gotoOperation.Target;
+                return target.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:152]: i don't understand this.  you're returning the syntax of the symbol just to try to find the symbol again, just to get the symbols locations...

you have the symbol, and it's location to begin with, why are you round-tripping through everything multiple times?

File: src/Features/VisualBasic/Portable/GoToDefinition/VisualBasicGoToDefinitionSymbolService.vb [hunk 8]
```diff
@@ -48,6 +50,29 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.GoToDefinition
                 Return exitTarget.GetLastToken().Span.End
             End If
 
+            Dim labelTarget = TryGetLabelTarget(semanticModel, node)
+            If labelTarget IsNot Nothing Then
+                Return labelTarget.GetFirstToken().Span.Start
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/GoToDefinition/VisualBasicGoToDefinitionSymbolService.vb:55]:
```suggestion
                Return labelTarget.SpanStart
```

File: src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs [hunk 9]
```diff
@@ -67,6 +70,13 @@ protected override ISymbol FindRelatedExplicitlyDeclaredSymbol(ISymbol symbol, C
 
                         return symbol.Locations.FirstOrDefault()?.SourceSpan.Start ?? 0;
                     }
+
+                case SyntaxKind.GotoKeyword:
+                case SyntaxKind.DefaultKeyword:
+                case SyntaxKind.CaseKeyword:
+                    {
+                        return TryFindAccessibleLabel(node)?.SpanStart;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/GoToDefinition/CSharpGoToDefinitionSymbolService.cs:78]: inline?

File: src/Features/VisualBasic/Portable/GoToDefinition/VisualBasicGoToDefinitionSymbolService.vb [hunk 10]
```diff
@@ -48,6 +50,29 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.GoToDefinition
...
 
+            Dim labelTarget = TryGetLabelTarget(semanticModel, node)
+            If labelTarget IsNot Nothing Then
+                Return labelTarget.SpanStart
+            End If
+
+            Return Nothing
+        End Function
+
+        Private Shared Function TryGetLabelTarget(semanticModel As SemanticModel, node As SyntaxNode) As SyntaxNode
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/GoToDefinition/VisualBasicGoToDefinitionSymbolService.vb:61]: inline?

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**